### DISCUSSION TITLE
Add support  managed  IL static link library

### DISCUSSION THREAD

**Discussion by John0King** [dotnet/runtime#85380] [Ideas]
current dotnet only support  dynamic link library , but for some   util class   or   internal used library  ,  for a library author it must introduce a  nuget dependency ,    and  also  find  in this repo(eg.  dotnet/runtimes)  ,   there is a  `~/Commons`  folder  for  the BCL library to use,  I wonder   can  dotnet  also support  static linking ?

this is something that I thought
-  default is internal for  static library.
-  add `export [<assemblyName/alais>:]<Type> [as <NewTypeName>]`  expression  to export static link library class/method/fields etc  into current assembly. 
-  nuget support

**Comment by hez2010** [comment]
Static linking simply doesn't work in dotnet libraries because you must provide your class libraries as assemblies. 
But the final artifact of your project can be statically linked.
If you are looking for a managed variant of statically linking, you can checkout ILLink(Trimming)+SingleFile publish.
If you are looking for a native variant, then you can checkout https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/interop.md

**Reply by John0King** [reply]
> Static linking simply doesn't work in dotnet libraries because you must provide your class libraries as assemblies.

that's ture , but  this is a idea for dotnet that may be impliement in the future.

I hope can specify `<OutputType>ILStaticLibrary</OutputType>`  and the project produce a `xxxx.libil`
and for the implimentation,  the worset implimentation is simply pack the source code into a "zip" file, and better way it to pick the IL class into the assembly that reference this "ILStaticLibrary" , the different from `ILLIink` is that  all the public  type in static link library should  modify to internal  and then merge to the "assembly" that reference it ( you can then export in the "assembly" and then the type will be a public type in "assembly")

**Reply by jkotas** [reply]
If you would like to, it is possible to package your component as sources. There are a few packages like that out there. For example, https://www.nuget.org/packages/xunit.assert.source . It is a bit involved to set it up, but it is certainly possible.
